src/222/main/kotlin/org/rust/cargo/project/model/impl/CargoSettingsFilesService.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.project.model.impl
+
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VfsUtil
+import com.intellij.util.io.systemIndependentPath
+import org.rust.cargo.CargoConstants
+import org.rust.cargo.project.model.CargoProject
+import org.rust.cargo.project.model.cargoProjects
+import org.rust.lang.RsFileType
+
+@Service
+class CargoSettingsFilesService(private val project: Project) {
+
+    @Volatile
+    private var settingsFilesCache: Map<String, SettingFileType>? = null
+
+    fun collectSettingsFiles(useCache: Boolean): Map<String, SettingFileType> {
+        return if (useCache) {
+            settingsFilesCache ?: collectSettingsFiles()
+        } else {
+            collectSettingsFiles()
+        }
+    }
+
+    private fun collectSettingsFiles(): Map<String, SettingFileType> {
+        val result = mutableMapOf<String, SettingFileType>()
+        for (cargoProject in project.cargoProjects.allProjects) {
+            result += cargoProject.collectSettingsFiles()
+        }
+
+        settingsFilesCache = result
+
+        return result
+    }
+
+    private fun CargoProject.collectSettingsFiles(): Map<String, SettingFileType> {
+        val files = mutableMapOf(
+            manifest.systemIndependentPath to SettingFileType.CONFIG,
+            manifest.parent.resolve(CargoConstants.LOCK_FILE).systemIndependentPath to SettingFileType.CONFIG
+        )
+
+        files += IMPLICIT_TARGET_FILES.mapNotNull {
+            val path = rootDir?.findFileByRelativePath(it)?.path ?: return@mapNotNull null
+            path to SettingFileType.IMPLICIT_TARGET
+        }
+
+        files += IMPLICIT_TARGET_DIRS
+            .mapNotNull { rootDir?.findFileByRelativePath(it) }
+            .flatMap { VfsUtil.collectChildrenRecursively(it) }
+            .mapNotNull { if (it.fileType == RsFileType) it.path to SettingFileType.IMPLICIT_TARGET else null }
```

Comment:
We're also interested in `rust-toolchain.toml` changes

Reply:
> It looks like we don't take into account Cargo.toml files (and all other "interesting" files) for all workspace crates

It seems so. Nice catch 👍 

> We're also interested in .cargo/config.toml changes
> "build.rs" is not a magic name, it's actually a custom name explicitly stated in Cargo.toml. Also, it can include submodules
> We're also interested in rust-toolchain.toml changes

This PR is not intended to cover any new case in comparison with `CargoTomlWatcher`. It's just about using new API
so `.cargo/config.toml` and `rust-toolchain.toml` will be covered in separate PR. See #6424 as a current plan what we need to do to turn this approach on and what I think should be enough to turn on build script evaluation after it

> Do you have an idea how we will track changes in proc macro crates using this API?

Since any change in proc macro lib requires updating of a project model, we can just consider any `rs` file in the corresponding crates as settings file. But I'm going to implement the same thing for build scripts first